### PR TITLE
[NFM] Problem: Missing chainmaind CLI level NFT types params validation and amino JSON support

### LINKED ISSUE #224: Problem: Introduce chainmaind CLI level NFT types params validation
The NFT module is missing validation for the params used.

### LINKED ISSUE #259: Problem: Missing NFT amino JSON support

### PR DESCRIPTION
**Pull request by calvinaco**
Fix #224
Fix #259

### CODE REVIEW

**Review by eddycpc — APPROVED** [review]

**Comment by cdc-Hitesh** [comment]
@calvinaco 
This is a precise piece of work. Awesome.
I have a small comment on the validation part:
- I see there is a URI length check on Golang CLI ([link](https://github.com/crypto-org-chain/chain-main/blob/88d06257ecd3a7a5c5cdf6dee0d4135b2661a6c9/x/nft/types/validation.go#L59))
- Also name cannot have a `space` in it (link)[https://github.com/crypto-org-chain/chain-main/blob/88d06257ecd3a7a5c5cdf6dee0d4135b2661a6c9/x/nft/types/validation.go#L39]

Do we wish to incorporate these changes as well? or we can go ahead without this too?

File: lib/src/cosmos/amino/msg.ts [hunk 1]
```diff
@@ -336,3 +336,88 @@ export function isMsgUndelegate(msg: Msg): msg is MsgUndelegate {
...
+        readonly data: string;
+        /** Bech32 encoded sender address */
+        readonly sender: string;
+        /** Bech32 encoded recipient address */
+        readonly recipient: string;
+    };
+}
+
+export function isMsgMintNFT(msg: Msg): msg is MsgMintNFT {
+    return (msg as MsgMintNFT).type === 'cosmos-sdk/MsgMintNFT';
```
Comment (cdc-Hitesh) [inline lib/src/cosmos/amino/msg.ts:372]: After testing I discovered the type will be `chainmain/nft/MsgMintNFT`

It doesn't accept `'cosmos-sdk/MsgMintNFT`

The same holds true for all NFT messages.

File: lib/src/cosmos/amino/msg.ts [hunk 2]
```diff
@@ -336,3 +336,88 @@ export function isMsgUndelegate(msg: Msg): msg is MsgUndelegate {
...
+        readonly name: string;
+        readonly uri: string;
+        readonly data: string;
+        /** Bech32 encoded sender address */
+        readonly sender: string;
+    };
+}
+
+export function isMsgEditNFT(msg: Msg): msg is MsgEditNFT {
+    return (msg as MsgEditNFT).type === 'cosmos-sdk/MsgEditNFT';
```
Comment (cdc-Hitesh) [inline lib/src/cosmos/amino/msg.ts:390]: Same as above, will need to change the `type` here

File: lib/src/cosmos/amino/msg.ts [hunk 3]
```diff
@@ -336,3 +336,88 @@ export function isMsgUndelegate(msg: Msg): msg is MsgUndelegate {
...
+        /** Token id */
+        readonly id: string;
+        readonly denom_id: string;
+        /** Bech32 encoded sender address */
+        readonly sender: string;
+    };
+}
+
+export function isMsgBurnNFT(msg: Msg): msg is MsgBurnNFT {
+    return (msg as MsgBurnNFT).type === 'cosmos-sdk/MsgBurnNFT';
```
Comment (cdc-Hitesh) [inline lib/src/cosmos/amino/msg.ts:405]: Same as above, will need to change the `type` here

File: lib/src/cosmos/amino/msg.ts [hunk 4]
```diff
@@ -336,3 +336,88 @@ export function isMsgUndelegate(msg: Msg): msg is MsgUndelegate {
...
+        readonly denom_id: string;
+        /** Bech32 encoded sender address */
+        readonly sender: string;
+        /** Bech32 encoded recipient address */
+        readonly recipient: string;
+    };
+}
+
+export function isMsgTransferNFT(msg: Msg): msg is MsgTransferNFT {
+    return (msg as MsgTransferNFT).type === 'cosmos-sdk/MsgTransferNFT';
```
Comment (cdc-Hitesh) [inline lib/src/cosmos/amino/msg.ts:422]: Same as above will need to change the `type` here.
For brevity I am skipping details

**Review by cdc-Hitesh — CHANGES_REQUESTED** [review]
The message type value needs to change.
I will create a separate PR as @calvinaco is on holiday today.

**Comment by cdc-Hitesh** [comment]
Marking above PR for NFM.

**Comment by calvinaco** [comment]
Close this in favour of #268

**Pull request closed without merging**